File: vapor-app/src/main/java/com/tevinjeffrey/vapor/okcloudapp/CloudAppService.java
package com.tevinjeffrey.vapor.okcloudapp;

import com.squareup.okhttp.RequestBody;
import com.tevinjeffrey.vapor.okcloudapp.model.AccountModel;
import com.tevinjeffrey.vapor.okcloudapp.model.AccountStatsModel;
import com.tevinjeffrey.vapor.okcloudapp.model.CloudAppItem;
import com.tevinjeffrey.vapor.okcloudapp.model.CloudAppJsonAccount;
import com.tevinjeffrey.vapor.okcloudapp.model.CloudAppJsonItem;
import com.tevinjeffrey.vapor.okcloudapp.model.ItemModel;
import com.tevinjeffrey.vapor.okcloudapp.model.UploadModel;

import java.util.List;
import java.util.Map;

import retrofit.Call;
import retrofit.http.Body;
import retrofit.http.DELETE;
import retrofit.http.GET;
import retrofit.http.Headers;
import retrofit.http.Multipart;
import retrofit.http.POST;
import retrofit.http.PUT;
import retrofit.http.Part;
import retrofit.http.PartMap;
import retrofit.http.Path;
import retrofit.http.QueryMap;
import rx.Observable;

public interface CloudAppService {
    @Headers("Accept: application/json")
    @GET("/{item-id}")
    Observable<ItemModel> getItem(@Path("item-id") String itemId);

    @Headers("Accept: application/json")
    @GET("/items")
    Observable<List<ItemModel>> listItems(@QueryMap Map<String, String> options);

    @Headers("Accept: application/json")
    @GET("/items/new")
    Observable<UploadModel> newUpload();

    @Headers("Accept: application/json")
    @DELETE("/items/{item-id}")
    Observable<ItemModel> deleteItem(@Path("item-id") String itemId);

    @Headers("Accept: application/json")
    @PUT("/items/{item-id}")
    Observable<ItemModel> renameItem(@Path("item-id") String itemId, @Body CloudAppJsonItem item);

    @Headers("Accept: application/json")
    @PUT("/items/{item-id}")
    Observable<ItemModel> setItemSecurity(@Path("item-id") String itemId, @Body CloudAppJsonItem item);

    @Headers("Accept: application/json")
    @POST("/items/{item-id}")
    Observable<ItemModel> bookmarkLink(@Path("item-id") String itemId, @Body CloudAppJsonItem item);

    @Headers("Accept: application/json")
    @GET("/account")
    Observable<AccountModel> getAccount();

    @Headers("Accept: application/json")
    @PUT("/account")
    Observable<AccountModel> changeDefaultSecurity(@Body CloudAppJsonAccount body);

    @Headers("Accept: application/json")
    @PUT("/account")
    Observable<AccountModel> changeEmail(@Body CloudAppJsonAccount body);

    @Headers("Accept: application/json")
    @POST("/account")
    Observable<AccountModel> changePassword(@Body CloudAppJsonAccount body);

    @Headers("Accept: application/json")
    @POST("/account")
    Observable<AccountModel> setCustomDomain(@Body CloudAppJsonAccount body);

    @Headers("Accept: application/json")
    @POST("/reset")
    Observable<AccountModel> resetPassword(@Body CloudAppJsonAccount body);

    @Headers("Accept: application/json")
    @POST("/register")
    Observable<AccountModel> registerAccount(@Body CloudAppJsonAccount body);

    @Headers("Accept: application/json")
    @GET("/account/stats")
    Observable<AccountStatsModel> getAccountStats();

    @Headers("Accept: application/json")
    @Multipart
    @POST("http://f.cl.ly")
    Observable<CloudAppItem> uploadFile(@PartMap Map<String, String> options, @Part("file") RequestBody filePart);
}

File: vapor-app/src/main/java/com/tevinjeffrey/vapor/okcloudapp/CloudAppService.kt
package com.tevinjeffrey.vapor.okcloudapp

import com.squareup.okhttp.RequestBody
import com.tevinjeffrey.vapor.okcloudapp.model.AccountModel
import com.tevinjeffrey.vapor.okcloudapp.model.AccountStatsModel
import com.tevinjeffrey.vapor.okcloudapp.model.CloudAppJsonAccount
import com.tevinjeffrey.vapor.okcloudapp.model.CloudAppJsonItem
import com.tevinjeffrey.vapor.okcloudapp.model.ItemModel
import com.tevinjeffrey.vapor.okcloudapp.model.UploadModel

import retrofit.http.Body
import retrofit.http.DELETE
import retrofit.http.GET
import retrofit.http.Headers
import retrofit.http.Multipart
import retrofit.http.POST
import retrofit.http.PUT
import retrofit.http.PartMap
import retrofit.http.Path
import retrofit.http.QueryMap
import rx.Observable

interface CloudAppService {
    @Headers("Accept: application/json")
    @GET("/{item-id}")
    fun getItem(@Path("item-id") itemId: String): Observable<ItemModel>

    @Headers("Accept: application/json")
    @GET("/items")
    fun listItems(@QueryMap options: Map<String, String>): Observable<List<ItemModel>>

    @Headers("Accept: application/json")
    @GET("/items/new")
    fun newUpload(@QueryMap parts: Map<String, String>): Observable<UploadModel>

    @Headers("Accept: application/json")
    @DELETE("/items/{item-id}")
    fun deleteItem(@Path("item-id") itemId: String): Observable<ItemModel>

    @Headers("Accept: application/json")
    @PUT("/items/{item-id}")
    fun renameItem(@Path("item-id") itemId: String, @Body item: CloudAppJsonItem): Observable<ItemModel>

    @Headers("Accept: application/json")
    @PUT("/items/{item-id}")
    fun setItemSecurity(@Path("item-id") itemId: String, @Body item: CloudAppJsonItem): Observable<ItemModel>

    @Headers("Accept: application/json")
    @POST("/items")
    fun bookmarkLink(@Body item: CloudAppJsonItem): Observable<ItemModel>

    @Headers("Accept: application/json")
    @GET("/account")
    fun getAccount(): Observable<AccountModel>

    @Headers("Accept: application/json")
    @PUT("/account")
    fun changeDefaultSecurity(@Body body: CloudAppJsonAccount): Observable<AccountModel>

    @Headers("Accept: application/json")
    @PUT("/account")
    fun changeEmail(@Body body: CloudAppJsonAccount): Observable<AccountModel>

    @Headers("Accept: application/json")
    @POST("/account")
    fun changePassword(@Body body: CloudAppJsonAccount): Observable<AccountModel>

    @Headers("Accept: application/json")
    @POST("/account")
    fun setCustomDomain(@Body body: CloudAppJsonAccount): Observable<AccountModel>

    @Headers("Accept: application/json")
    @POST("/reset")
    fun resetPassword(@Body body: CloudAppJsonAccount): Observable<AccountModel>

    @Headers("Accept: application/json")
    @POST("/register")
    fun registerAccount(@Body body: CloudAppJsonAccount): Observable<AccountModel>

    @Headers("Accept: application/json")
    @GET("/account/stats")
    fun getAccountStats(): Observable<AccountStatsModel>

    @Multipart
    @Headers("Accept: application/json")
    @POST("http://s3.amazonaws.com/f.cl.ly")
    fun uploadFile(@PartMap parts: Map<String, RequestBody>): Observable<ItemModel>
}
